kafka/consumer.go

```diff
@@ -153,39 +174,80 @@ func NewConsumer(cfg ConsumerConfig) (*Consumer, error) {
 		cfg.MaxPollRecords = 100
 	}
 	return &Consumer{
-		cfg:      cfg,
-		client:   client,
-		consumer: consumer,
-		tracer:   cfg.tracerProvider().Tracer("kafka"),
+		cfg:        cfg,
+		client:     client,
+		consumer:   consumer,
+		closed:     make(chan struct{}),
+		running:    make(chan struct{}),
+		forceClose: forceClose,
+		stopPoll:   func() {},
+		tracer:     cfg.tracerProvider().Tracer("kafka"),
 	}, nil
 }
 
 // Close the consumer, blocking until all partition consumers are stopped.
 func (c *Consumer) Close() error {
-	if !c.closed.CompareAndSwap(false, true) {
-		return nil
-	}
 	c.mu.Lock()
 	defer c.mu.Unlock()
-	c.client.Close()
-	c.consumer.wg.Wait() // Wait for all the goroutines to exit.
+	select {
+	case <-c.closed:
+	default:
+		close(c.closed)
+		// Cancel the context used in client.PollRecords, triggering graceful
+		// cancellation.
+		c.stopPoll()
+		stopped := make(chan struct{})
+		go func() {
+			defer close(stopped)
+			// Close all partition consumers first to ensure there aren't any
+			// records being processed while the kgo.Client is being closed.
+			// Also ensures that commits can be issued after the records are
+			// processed when AtLeastOnceDelivery is configured.
+			c.consumer.close()
+		}()
+		// Wait for the consumers to process any in-flight records, or cancel
+		// the underlying processing context if they aren't stopped in time.
+		select {
+		case <-time.After(c.cfg.ShutdownGracePeriod): // Timeout
+			c.forceClose(fmt.Errorf(
+				"consumer: close: timeout waiting for consumers to stop (%s)",
+				c.cfg.ShutdownGracePeriod.String(),
+			))
+		case <-stopped: // Stopped within c.cfg.ShutdownGracePeriod.
+		}
+		c.client.Close() // Last, close the `kgo.Client`
```

Comment:
Should we do this as a defer, after the `close(c.closed)`? Or does it need to happen before we wait for the waitgroup?

Reply:
I suppose it can be done as a defer, but it seemed more appropriate to keep it outside of the defer just to be able to visualize the order of calls more easily. I can add it in a defer after `close(c.closed)` if you think that's better.

Reply:
I'm mildly concerned that we might update the code and introduce a code path that would lead to the kgo.Client being left open. I'll leave it to your judgement, just a passing thought.